PumaRhino/PumaGrasshopperAPI.cpp

```diff
@@ -93,186 +102,351 @@ inline RHINOPRT_API bool AddInitialMesh(ON_SimpleArray<const ON_Mesh*>* pMesh) {
+
+		indexStartBool += boolAttrCount;
+		indexStartDouble += doubleAttrCount;
+		indexStartString += stringAttrCount;
+		indexStartBoolArray += boolAttrArrayCount;
+		indexStartDoubleArray += doubleAttrArrayCount;
+		indexStartStringArray += stringAttrArrayCount;
+	}
 
-RHINOPRT_API bool Generate(ON_SimpleArray<int>* pMeshCounts, ON_SimpleArray<ON_Mesh*>* pMeshArray) {
-	RhinoPRT::get().GenerateGeometry();
-	const std::vector<GeneratedModelPtr>& models = RhinoPRT::get().getGenModels();
+	const auto& models = RhinoPRT::get().GenerateGeometry(std::wstring(rpk_path), rawInitialShapes, aBuilders);
 
-	for (size_t initialShapeIndex = 0; initialShapeIndex < models.size(); initialShapeIndex++) {
-		if (models[initialShapeIndex]) {
+	for (size_t i = 0; i < models.size(); i++) {
+		if (models[i]) {
 			const GeneratedModel::MeshBundle meshBundle =
-			        models[initialShapeIndex]->createRhinoMeshes(initialShapeIndex);
+			        models[i]->createRhinoMeshes(i);
 			pMeshCounts->Append(static_cast<int>(meshBundle.size()));
 			for (const auto& meshPart : meshBundle) {
 				pMeshArray->Append(new ON_Mesh(meshPart));
 			}
-		}
-		else {
-			pMeshCounts->Append(0);
-		}
-	}
-
-	return !models.empty();
-}
 
-RHINOPRT_API int GetMeshPartCount(int initialShapeIndex) {
-	const auto& models = RhinoPRT::get().getGenModels();
+			// Materials
+			pMatIndices->Append(meshBundle.size());
 
-	if (models.size() <= initialShapeIndex || !models[initialShapeIndex])
-		return 0;
+			const auto& materials = models[i]->getMaterials();
+			for (const auto& material : materials) {
+				const auto& matAttributes = material.second;
+				pcu::appendColor(matAttributes.mDiffuseCol, pColorsArray);
+				pcu::appendColor(matAttributes.mAmbientCol, pColorsArray);
+				pcu::appendColor(matAttributes.mSpecularCol, pColorsArray);
+				pColorsArray->Append(matAttributes.mOpacity);
+				pColorsArray->Append(matAttributes.mShininess);
 
-	return models[initialShapeIndex]->getMeshPartCount();
-}
-
-RHINOPRT_API int GetRuleAttributesCount() {
-	return RhinoPRT::get().GetRuleAttributeCount();
-}
+				pMatIndices->Append(matAttributes.mTexturePaths.size());
 
-RHINOPRT_API bool GetRuleAttribute(int attrIdx, ON_wString* pRule, ON_wString* pName, ON_wString* pNickname,
-                                   prt::AnnotationArgumentType* type, ON_wString* pGroup) {
-	const RuleAttributes& ruleAttributes = RhinoPRT::get().GetRuleAttributes();
+				for (auto& texture : matAttributes.mTexturePaths) {
 
-	if (attrIdx >= ruleAttributes.size())
-		return false;
-
-	const RuleAttributeUPtr& ruleAttr = ruleAttributes[attrIdx];
-	pcu::appendToRhinoString(*pRule, ruleAttr->mRuleFile);
-	pcu::appendToRhinoString(*pName, ruleAttr->mFullName);
-	pcu::appendToRhinoString(*pNickname, ruleAttr->mNickname);
-	*type = ruleAttr->mType;
-
-	if (ruleAttr->groups.size() > 0)
-		*pGroup += ON_wString(ruleAttr->groups.front().c_str());
-
-	return true;
-}
+#ifdef DEBUG
```

Comment:
Consider introducing a `constexpr bool DBG` and use `if constexpr (DBG)` to query it.
The value of DBG can still be controlled through DEBUG compile definition if needed.